tls/extensions/s2n_server_certificate_request_extensions.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "error/s2n_errno.h"
+#include "s2n_server_signature_algorithms.h"
+#include "utils/s2n_safety.h"
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls13.h"
+
+int s2n_server_certificate_request_extensions_parse(struct s2n_connection *conn, struct s2n_blob *extensions)
+{
+    bool processed_sig_algs = 0;
+    struct s2n_stuffer ext_stuffer = {0};
+    GUARD(s2n_stuffer_init(&ext_stuffer, extensions));
+    GUARD(s2n_stuffer_write(&ext_stuffer, extensions));
+
+    while (s2n_stuffer_data_available(&ext_stuffer)) {
+        struct s2n_blob ext = {0};
+        uint16_t extension_type, extension_size;
+        struct s2n_stuffer current_extension = {0};
+
+        /* check there is at least length 4 available to read the following two uint16 values */
+        S2N_ERROR_IF(s2n_stuffer_data_available(&ext_stuffer) < 4, S2N_ERR_BAD_MESSAGE);
+        GUARD(s2n_stuffer_read_uint16(&ext_stuffer, &extension_type));
+        GUARD(s2n_stuffer_read_uint16(&ext_stuffer, &extension_size));
+
+        ext.size = extension_size;
+        ext.data = s2n_stuffer_raw_read(&ext_stuffer, ext.size);
+        notnull_check(ext.data);
+
+        GUARD(s2n_stuffer_init(&current_extension, &ext));
+        GUARD(s2n_stuffer_write(&current_extension, &ext));
+
+        switch (extension_type) {
+        case TLS_EXTENSION_SIGNATURE_ALGORITHMS:
+            GUARD(s2n_extensions_server_signature_algorithms_recv(conn, &current_extension));
```

Comment:
do we need to guard against duplicate extensions?

Reply:
yes. there is a ongoing parallel task to redo how we do our extensions. Currently that is a piece missing across all extension parsing. I wrote it to match how the other server extensions are parsed so they would be updated uniformly. But it would not be difficult to add in

Reply:
I think we do handle duplicate extensions in our other extension code paths already? For example, this looks like we parse before processing which gives us the ability to easily filter duplicates before calling any of the extension functions: https://github.com/awslabs/s2n/blob/6e25fbaf99ad4e117127c87c400ac0c4e4f97053/tls/s2n_client_hello.c#L258